src/Features/Core/Portable/EventListener/EventListenerTracker.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.EventListener
+{
+    internal class EventListenerTracker<TService>
+    {
+        private readonly HashSet<string> _eventListenerInitialized;
+        private readonly IEnumerable<Lazy<IEventListener, EventListenerMetadata>> _eventListeners;
+
+        public EventListenerTracker(IEnumerable<Lazy<IEventListener, EventListenerMetadata>> eventListeners)
+        {
+            _eventListenerInitialized = new HashSet<string>();
+            _eventListeners = eventListeners;
+        }
+
+        public void EnsureEventListener(Workspace workspace, TService service)
+        {
+            lock (_eventListenerInitialized)
+            {
+                if (!_eventListenerInitialized.Add(workspace.Kind))
```

Comment:
What happens if we have more than one instance of the same kind of workspace? Think if we try leveraging this for the preview workspace or something.

Reply:
that is not supported scenario. this only was ensure, but it doesn't have cleanup. meaning it currently expects once it happens it is there until VS goes away.

we can try to support such case by holding onto Worksapce and ConditionalWeakTable or have explicit Remove call, but better to add such support when we actually have a use case for it

Reply:
current helper doesn't support the case you mentioned. it expects event listener to have same lifetime as service. so there is no stop in event listener.